File: updater.js
/**
 * @fileoverview GitHub API 기반 자동 업데이트 모듈
 * electron-updater 대신 GitHub API로 직접 버전 확인
 * @module updater
 */

const { ipcMain, shell } = require('electron');
const https = require('https');
const { version } = require('./package.json');
const logger = require('./logger');

/** @type {string} GitHub 저장소 소유자 */
const GITHUB_OWNER = 'minseok7891';

/** @type {string} GitHub 저장소 이름 */
const GITHUB_REPO = 'fazzk';

/**
 * 메인 윈도우 참조
 * @type {Electron.BrowserWindow|null}
 */
let mainWindow = null;

/**
 * 업데이터 초기화
 * @param {Electron.BrowserWindow} win - 메인 윈도우 인스턴스
 * @returns {void}
 */
function initUpdater(win) {
    mainWindow = win;
}

/**
 * GitHub에서 최신 릴리즈 정보 가져오기
 * @async
 * @returns {Promise<Object>} GitHub 릴리즈 객체
 */
function getLatestRelease() {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: 'api.github.com',
            path: `/repos/${GITHUB_OWNER}/${GITHUB_REPO}/releases/latest`,
            method: 'GET',
            headers: {
                'User-Agent': 'Fazzk-Updater',
                'Accept': 'application/vnd.github.v3+json'
            }
        };

        const req = https.request(options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const release = JSON.parse(data);
                    resolve(release);
                } catch (e) {
                    reject(e);
                }
            });
        });

        req.on('error', reject);
        req.end();
    });
}

/**
 * Semantic Version 비교
 * @param {string} v1 - 첫 번째 버전 (예: "v1.2.3")
 * @param {string} v2 - 두 번째 버전 (예: "v1.2.4")
 * @returns {number} v1 > v2이면 1, v1 < v2이면 -1, 같으면 0
 */
function compareVersions(v1, v2) {
    const parts1 = v1.replace('v', '').split('.').map(Number);
    const parts2 = v2.replace('v', '').split('.').map(Number);

    for (let i = 0; i < 3; i++) {
        if (parts1[i] > parts2[i]) return 1;
        if (parts1[i] < parts2[i]) return -1;
    }
    return 0;
}

/**
 * @typedef {Object} UpdateCheckResult
 * @property {boolean} hasUpdate - 업데이트 가능 여부
 * @property {string} [latestVersion] - 최신 버전 (업데이트 있을 시)
 * @property {string} [downloadUrl] - 다운로드 URL (업데이트 있을 시)
 * @property {string} [error] - 에러 메시지 (실패 시)
 */

/**
 * 업데이트 확인
 * @async
 * @returns {Promise<UpdateCheckResult>} 업데이트 확인 결과
 */
async function checkForUpdates() {
    try {
        const release = await getLatestRelease();
        const latestVersion = release.tag_name;
        const currentVersion = `v${version}`;

        if (compareVersions(latestVersion, currentVersion) > 0) {
            const exeAsset = release.assets.find(a => a.name.endsWith('.exe'));

            if (mainWindow) {
                mainWindow.webContents.send('update-available-github', {
                    currentVersion: currentVersion,
                    latestVersion: latestVersion,
                    releaseNotes: release.body || '',
                    downloadUrl: exeAsset ? exeAsset.browser_download_url : release.html_url,
                    releaseUrl: release.html_url,
                    publishedAt: release.published_at
                });
            }

            return {
                hasUpdate: true,
                latestVersion,
                downloadUrl: exeAsset ? exeAsset.browser_download_url : release.html_url
            };
        }

        return { hasUpdate: false };
    } catch (error) {
        logger.error('[Updater] 업데이트 확인 실패:', error.message);
        return { hasUpdate: false, error: error.message };
    }
}

/**
 * 외부 브라우저에서 다운로드 페이지 열기
 * @param {string} url - 열 URL
 * @returns {void}
 */
function openDownloadPage(url) {
    shell.openExternal(url);
}

// IPC 핸들러
ipcMain.handle('check-for-updates', async () => {
    return await checkForUpdates();
});

ipcMain.handle('open-download-page', (_, url) => {
    openDownloadPage(url);
});

module.exports = {
    initUpdater,
    checkForUpdates
};

File: src-tauri/src/updater.rs
//! GitHub API 기반 자동 업데이트 모듈
//!
//! GitHub Releases에서 최신 버전을 확인하고 업데이트 알림을 제공합니다.

use reqwest::header::{HeaderMap, HeaderValue, USER_AGENT};
use serde::{Deserialize, Serialize};
use std::sync::OnceLock;

/// 전역 HTTP 클라이언트 (재사용)
static HTTP_CLIENT: OnceLock<reqwest::Client> = OnceLock::new();

/// HTTP 클라이언트 가져오기
fn get_http_client() -> &'static reqwest::Client {
    HTTP_CLIENT.get_or_init(|| reqwest::Client::new())
}

/// GitHub 저장소 정보
const GITHUB_OWNER: &str = "minseok7891";
const GITHUB_REPO: &str = "fazzk";

/// 현재 앱 버전 (Cargo.toml에서 가져옴)
const CURRENT_VERSION: &str = env!("CARGO_PKG_VERSION");

/// GitHub Release 응답 구조체
#[derive(Debug, Deserialize)]
struct GitHubRelease {
    tag_name: String,
    html_url: String,
    body: Option<String>,
    published_at: Option<String>,
    assets: Vec<GitHubAsset>,
}

/// GitHub Asset 구조체
#[derive(Debug, Deserialize)]
struct GitHubAsset {
    name: String,
    browser_download_url: String,
}

/// 업데이트 확인 결과
#[derive(Debug, Serialize)]
pub struct UpdateCheckResult {
    pub has_update: bool,
    pub current_version: String,
    pub latest_version: Option<String>,
    pub download_url: Option<String>,
    pub release_url: Option<String>,
    pub release_notes: Option<String>,
    pub published_at: Option<String>,
    pub error: Option<String>,
}

/// Semantic Version 비교
///
/// v1 > v2 이면 1, v1 < v2 이면 -1, 같으면 0
fn compare_versions(v1: &str, v2: &str) -> i32 {
    let parse = |v: &str| -> Vec<u32> {
        v.trim_start_matches('v')
            .split('.')
            .filter_map(|s| s.parse().ok())
            .collect()
    };

    let parts1 = parse(v1);
    let parts2 = parse(v2);

    for i in 0..3 {
        let p1 = parts1.get(i).copied().unwrap_or(0);
        let p2 = parts2.get(i).copied().unwrap_or(0);

        if p1 > p2 {
            return 1;
        }
        if p1 < p2 {
            return -1;
        }
    }
    0
}

/// GitHub에서 최신 릴리즈 정보 가져오기
async fn get_latest_release() -> Result<GitHubRelease, String> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/releases/latest",
        GITHUB_OWNER, GITHUB_REPO
    );

    let mut headers = HeaderMap::new();
    headers.insert(USER_AGENT, HeaderValue::from_static("Fazzk-Tauri-Updater"));
    headers.insert(
        "Accept",
        HeaderValue::from_static("application/vnd.github.v3+json"),
    );

    let client = get_http_client();
    let res = client
        .get(&url)
        .headers(headers)
        .send()
        .await
        .map_err(|e| format!("네트워크 오류: {}", e))?;

    if !res.status().is_success() {
        return Err(format!("GitHub API 오류: {}", res.status()));
    }

    res.json::<GitHubRelease>()
        .await
        .map_err(|e| format!("응답 파싱 오류: {}", e))
}

/// 업데이트 확인 (Tauri Command)
#[tauri::command]
pub async fn check_for_updates() -> UpdateCheckResult {
    let current_version = format!("v{}", CURRENT_VERSION);

    match get_latest_release().await {
        Ok(release) => {
            let latest_version = release.tag_name.clone();

            if compare_versions(&latest_version, &current_version) > 0 {
                // 업데이트 있음
                let download_url = release
                    .assets
                    .iter()
                    .find(|a| a.name.ends_with(".exe") || a.name.ends_with(".msi"))
                    .map(|a| a.browser_download_url.clone())
                    .unwrap_or_else(|| release.html_url.clone());

                println!(
                    "[Updater] 업데이트 발견: {} -> {}",
                    current_version, latest_version
                );

                UpdateCheckResult {
                    has_update: true,
                    current_version,
                    latest_version: Some(latest_version),
                    download_url: Some(download_url),
                    release_url: Some(release.html_url),
                    release_notes: release.body,
                    published_at: release.published_at,
                    error: None,
                }
            } else {
                // 최신 버전 사용 중
                println!("[Updater] 최신 버전 사용 중: {}", current_version);

                UpdateCheckResult {
                    has_update: false,
                    current_version,
                    latest_version: Some(latest_version),
                    download_url: None,
                    release_url: None,
                    release_notes: None,
                    published_at: None,
                    error: None,
                }
            }
        }
        Err(e) => {
            eprintln!("[Updater] 업데이트 확인 실패: {}", e);

            UpdateCheckResult {
                has_update: false,
                current_version,
                latest_version: None,
                download_url: None,
                release_url: None,
                release_notes: None,
                published_at: None,
                error: Some(e),
            }
        }
    }
}

/// 다운로드 페이지 열기 (Tauri Command)
#[tauri::command]
pub async fn open_download_page(url: String) -> Result<(), String> {
    #[cfg(target_os = "windows")]
    std::process::Command::new("cmd")
        .args(["/C", "start", &url])
        .spawn()
        .map_err(|e| format!("브라우저 열기 실패: {}", e))?;

    #[cfg(not(target_os = "windows"))]
    std::process::Command::new("xdg-open")
        .arg(&url)
        .spawn()
        .map_err(|e| format!("브라우저 열기 실패: {}", e))?;

    Ok(())
}

use tauri::{AppHandle, Emitter};

#[derive(Clone, Serialize)]
struct ProgressPayload {
    percent: u64,
    total: u64,
    current: u64,
}

/// 앱 내에서 업데이트 직접 다운로드 및 사일런트 설치
#[tauri::command]
pub async fn download_and_install_update(app: AppHandle, url: String) -> Result<(), String> {
    println!("[Updater] 다운로드 시작: {}", url);

    // 1. Temp 파일 생성
    let temp_dir = std::env::temp_dir();
    let file_name = url.split('/').next_back().unwrap_or("fazzk_update.exe");
    let file_path = temp_dir.join(file_name);

    println!("[Updater] 저장 경로: {:?}", file_path);

    // 2. reqwest로 스트리밍 다운로드 (실시간 진행률 표시)
    let client = get_http_client();
    let res = client
        .get(&url)
        .header(USER_AGENT, "Fazzk-Updater")
        .send()
        .await
        .map_err(|e| format!("네트워크 요청 실패: {}", e))?;

    let total_size = res.content_length().unwrap_or(0);

    // 파일 생성
    let mut file =
        std::fs::File::create(&file_path).map_err(|e| format!("파일 생성 실패: {}", e))?;

    // 스트리밍 다운로드
    use futures_util::StreamExt;
    use std::io::Write;

    let mut stream = res.bytes_stream();
    let mut downloaded: u64 = 0;

    while let Some(chunk_result) = stream.next().await {
        let chunk = chunk_result.map_err(|e| format!("청크 다운로드 실패: {}", e))?;

        file.write_all(&chunk)
            .map_err(|e| format!("파일 쓰기 실패: {}", e))?;

        downloaded += chunk.len() as u64;

        // 진행률 계산 및 전송
        let percent = if total_size > 0 {
            (downloaded as f64 / total_size as f64 * 100.0) as u64
        } else {
            0
        };

        app.emit(
            "update-progress",
            ProgressPayload {
                percent,
                total: total_size,
                current: downloaded,
            },
        )
        .unwrap_or_default();
    }

    // 파일 닫기
    drop(file);

    println!("[Updater] 다운로드 완료, 1초 대기 후 사일런트 설치 시작...");

    // 100% 표시 후 1초 대기
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;

    // 3. 설치 후 자동 실행을 위한 배치 스크립트 생성
    let app_exe_path =
        std::env::current_exe().map_err(|e| format!("현재 실행 파일 경로 가져오기 실패: {}", e))?;

    let batch_script = format!(
        r#"@echo off
timeout /t 2 /nobreak > nul
"{}" /S
timeout /t 3 /nobreak > nul
start "" "{}"
(goto) 2>nul & del "%~f0"
exit
"#,
        file_path.display(),
        app_exe_path.display()
    );

    let batch_path = temp_dir.join("fazzk_update.bat");
    std::fs::write(&batch_path, batch_script)
        .map_err(|e| format!("배치 스크립트 생성 실패: {}", e))?;

    println!("[Updater] 배치 스크립트 생성: {:?}", batch_path);

    // 4. 배치 스크립트 실행 (완전히 숨김 - CREATE_NO_WINDOW)
    #[cfg(target_os = "windows")]
    {
        use std::os::windows::process::CommandExt;
        const CREATE_NO_WINDOW: u32 = 0x08000000;

        std::process::Command::new("cmd")
            .args(["/C", &batch_path.to_string_lossy()])
            .creation_flags(CREATE_NO_WINDOW)
            .spawn()
            .map_err(|e| format!("업데이트 스크립트 실행 실패: {}", e))?;
    }

    #[cfg(not(target_os = "windows"))]
    {
        std::process::Command::new("sh")
            .arg(&batch_path)
            .spawn()
            .map_err(|e| format!("업데이트 스크립트 실행 실패: {}", e))?;
    }

    // 5. 앱 종료 (설치 진행을 위해)
    std::process::exit(0);
}
